Do not load ECE button if total amount is 0

Mayisha opened the pull request:
## Changes proposed in this Pull Request:
- Not loading the ECE button when your cart total amount is `0`. Because ECE expects the amount to be greater than `0`.

- Fixed the following console error.
``` Uncaught Typeerror: Can not read properties of undefined (reading 'wcSettings')```


## Testing instructions
- Create a free product.
- Add the product to your cart and go to the cart/checkout page.
- In `develop` branch, you should see the following console error.

<img width="431" alt="Screenshot 2024-10-28 at 8 33 14 PM" src="https://github.com/user-attachments/assets/d23d3db9-0980-424a-bc0c-50abab88230e">

- In this branch, you should not see any console errors.

**Note**: the branch name is misleading as I initially targeted a different fix. The trial subscription will be fixed in a separate issue as it needs some more investigation (It's still under spike in WooPayments as well).

Mayisha committed:
hide ece button if total amount is 0

Mayisha committed:
Merge branch 'develop' into task/free-trial-sub

Mayisha committed:
update comment

Mayisha committed:
add changelog

annemirasol commented on includes/payment-methods/class-wc-stripe-express-checkout-helper.php:
```diff
@@ -543,6 +543,14 @@ public function should_show_express_checkout_button() {
 
+		// Don't show if the total price is 0.
+		// ToDo: support free trials. Free trials should be supported if the product does not require shipping.
+		if ( 0.0 === (float) WC()->cart->get_total( false )
+			|| ( $this->is_product() && 0.0 === (float) $this->get_product()->get_price() )
```
When my cart total is 0 (i.e. empty cart, free product, or with big enough coupon), and I visit the product page for a product that is not free, I think we still want to have express payment available. Same with Pay for Order page.

Something like?

```suggestion
		if ( ( ( is_cart() || is_checkout() ) && 0.0 === (float) WC()->cart->get_total( false ) )
			|| ( $this->is_product() && 0.0 === (float) $this->get_product()->get_price() )
```

Mayisha replied:
That's a good catch 👍

annemirasol reviewed (changes requested):
Thanks for fixing this! ✨  Found some corner cases involving product and pay for order pages, when cart is empty.

Mayisha committed:
check for cart and checkout page

Mayisha committed:
fix condition for pay for order page

Mayisha committed:
Merge branch 'develop' into task/free-trial-sub

Mayisha commented:
Thanks @annemirasol for the feedback.

I have addressed it in my latest commits. I have added a slightly different condition than the one you suggested. This is because `is_checkout` returns `true` on the pay for order page, making the button hidden wrongly.

annemirasol commented on client/blocks/utils.js:
```diff
@@ -1,7 +1,7 @@
 
 export const getBlocksConfiguration = () => {
-	const stripeServerData = getSetting( 'stripe_data', null );
+	const stripeServerData = wc?.wcSettings?.getSetting( 'stripe_data', null );
```
What is this changing?

Mayisha replied:
I was getting this ( p1730176005939309-slack-C05GDK92WGM ) console error which I fixed this way.

annemirasol replied:
Great, thanks for fixing that!

annemirasol reviewed (approved):
LGTM 🚢 

Fix works great ✨ I don't get any integration errors anymore when cart total is 0, and the product and pay-for-order pages are unaffected by cart state.

Non-blocking question about the `getSetting()` change.

Mayisha committed:
Merge branch 'develop' into task/free-trial-sub
